src/Events/QR/Redirections.php

```diff
@@ -0,0 +1,251 @@
+			'order'          => 'ASC',
+		];
+		$events = tribe_get_events( $args );
+
+		$url = empty( $events ) ? $this->get_fallback_url() : get_permalink( $events[0]->ID );
+
+		/**
+		 * Filters the URL for the upcoming event redirection.
+		 *
+		 * @since TBD
+		 *
+		 * @param string $url     The URL to redirect to.
+		 * @param array  $events  The events found by the query.
+		 * @param self   $context The Redirections instance.
+		 */
+		return apply_filters( 'tec_events_qr_upcoming_event_url', $url, $events, $this );
+	}
+
+	/**
+	 * Get the URL for a specific event.
+	 *
+	 * @since TBD
+	 * @param int $post_id The post ID of the event.
+	 * @return string The URL to redirect to, either an event permalink or fallback URL.
+	 */
+	public function get_specific_event_url( int $post_id ): string {
+		$post_type = get_post_type( $post_id );
+		if ( ! $post_type || TEC::POSTTYPE !== $post_type ) {
+			return $this->get_fallback_url();
+		}
+
+		$url = get_permalink( $post_id );
+
+		/**
+		 * Filters the URL for the specific event redirection.
+		 *
+		 * @since TBD
+		 *
+		 * @param string $url     The URL to redirect to.
+		 * @param int    $post_id The post ID of the event.
+		 * @param self   $context The Redirections instance.
+		 */
+		return apply_filters( 'tec_events_qr_specific_event_url', $url, $post_id, $this );
+	}
+
+	/**
+	 * Get the URL for the next event in a series.
+	 *
+	 * @since TBD
+	 * @param int $post_id The post ID of the series.
+	 * @return string The URL to redirect to, either an event permalink or fallback URL.
+	 */
+	public function get_next_series_event_url( int $post_id ): string {
+		// If we don't have the Pro version, return the fallback URL.
+		if ( ! has_action( 'tribe_common_loaded', 'tribe_register_pro' ) ) {
+			return $this->get_fallback_url();
+		}
+
+		// Get the next event in the series using the Series_Relationship class.
+		$next_event = \TEC\Events_Pro\Custom_Tables\V1\Models\Series_Relationship::next( $post_id );
+
+		$url = $next_event ? get_permalink( $next_event->post_id ) : $this->get_fallback_url();
+
+		/**
+		 * Filters the URL for the next event in a series redirection.
+		 *
+		 * @since TBD
+		 *
+		 * @param string $url     The URL to redirect to.
+		 * @param array  $events  The events found by the query.
+		 * @param int    $post_id The post ID of the series.
+		 * @param self   $context The Redirections instance.
+		 */
+		return apply_filters( 'tec_events_qr_next_series_event_url', $url, [ $next_event ], $post_id, $this );
+	}
+
+	/**
+	 * Handle QR code redirections.
+	 *
+	 * @since TBD
+	 * @return void
+	 */
+	public function handle_qr_redirect(): void {
+		$hash = get_query_var( 'tec_qr_hash' );
+
+		if ( ! $hash ) {
+			return;
+		}
+
+		$routes = tribe( Routes::class );
+
+		try {
+			$data = $routes->decode_qr_hash( $hash );
+		} catch ( \InvalidArgumentException $e ) {
+			wp_safe_redirect( esc_url( $this->get_fallback_url() ) );
+			exit;
```

Comment:
Replace exit with tribe_exit - you'll need to ignore the phpcs errors that come with this change